aspnetcore/blazor/advanced-scenarios.md

```diff
@@ -0,0 +1,163 @@
+
+A key example of these improvements involve *sequence numbers*. Sequence numbers indicate to the runtime which outputs came from which distinct and ordered lines of code. The runtime uses this information to generate efficient tree diffs in linear time, which is far faster than is normally possible for a general tree diff algorithm.
+
+Consider the following Razor component (*.razor*) file:
+
+```razor
+@if (someFlag)
+{
+    <text>First</text>
+}
+
+Second
+```
+
+The preceding code compiles to something like the following:
+
+```csharp
+if (someFlag)
+{
+    builder.AddContent(0, "First");
+}
+
+builder.AddContent(1, "Second");
+```
+
+When the code executes for the first time, if `someFlag` is `true`, the builder receives:
+
+| Sequence | Type      | Data   |
+| :------: | --------- | :----: |
+| 0        | Text node | First  |
+| 1        | Text node | Second |
+
+Imagine that `someFlag` becomes `false`, and the markup is rendered again. This time, the builder receives:
+
+| Sequence | Type       | Data   |
+| :------: | ---------- | :----: |
+| 1        | Text node  | Second |
+
+When the runtime performs a diff, it sees that the item at sequence `0` was removed, so it generates the following trivial *edit script*:
+
+* Remove the first text node.
+
+### What goes wrong if you generate sequence numbers programmatically
+
+Imagine instead that you wrote the following render tree builder logic:
+
+```csharp
+var seq = 0;
+
+if (someFlag)
+{
+    builder.AddContent(seq++, "First");
+}
+
+builder.AddContent(seq++, "Second");
+```
+
+Now, the first output is:
+
+| Sequence | Type      | Data   |
+| :------: | --------- | :----: |
+| 0        | Text node | First  |
+| 1        | Text node | Second |
+
+This outcome is identical to the prior case, so no negative issues exist. `someFlag` is `false` on the second rendering, and the output is:
+
+| Sequence | Type      | Data   |
+| :------: | --------- | ------ |
+| 0        | Text node | Second |
+
+This time, the diff algorithm sees that *two* changes have occurred, and the algorithm generates the following edit script:
+
+* Change the value of the first text node to `Second`.
+* Remove the second text node.
+
+Generating the sequence numbers has lost all the useful information about where the `if/else` branches and loops were present in the original code. This results in a diff **twice as long** as before.
+
+This is a trivial example. In more realistic cases with complex and deeply nested structures, and especially with loops, the performance cost is more severe. Instead of immediately identifying which loop blocks or branches have been inserted or removed, the diff algorithm has to recurse deeply into the render trees and usually build far longer edit scripts because it is misinformed about how the old and new structures relate to each other.
```

Comment:
```suggestion
This is a trivial example. In more realistic cases with complex and deeply nested structures, and especially with loops, the performance cost is more severe. Instead of immediately identifying which loop blocks or branches have been inserted or removed, the diff algorithm has to recurse deeply into the render trees and usually build longer edit scripts because it's misinformed about how the old and new structures relate to each other.
```

Also consider splitting the last sentence into 2